src/save.rs

```diff
@@ -0,0 +1,247 @@
+                    .context(WriteToTiffSnafu)?;
+
+                metadata
+                    .write_tags(&mut tiff)
+                    .map_err(|e| SaveError::WriteTags {
+                        source: Box::new(e),
+                    })?;
+                let bytes = image.$as_fn().ok_or(SaveError::ConvertImageToBytes)?;
+                tiff.write_data(bytes).context(WriteToTiffSnafu)?;
+                Ok(())
+            }
+        }
+    };
+}
+
+// Implementations for Gray16
+impl_save_frame!(Gray16, Uncompressed, as_luma16);
+impl_save_frame!(Gray16, Packbits, as_luma16);
+impl_save_frame!(Gray16, Lzw, as_luma16);
+impl_save_frame!(Gray16, Deflate, as_luma16);
+
+// Implementations for RGB8
+impl_save_frame!(RGB8, Uncompressed, as_rgb8);
+impl_save_frame!(RGB8, Packbits, as_rgb8);
+impl_save_frame!(RGB8, Lzw, as_rgb8);
+impl_save_frame!(RGB8, Deflate, as_rgb8);
+
+impl TiffSaver {
+    pub fn new(compressor: Compressor, color: DicomColorType) -> Self {
+        Self { compressor, color }
+    }
+
+    pub fn save(
+        &self,
+        encoder: &mut TiffEncoder<File>,
+        image: &DynamicImage,
+        metadata: &PreprocessingMetadata,
+    ) -> Result<(), SaveError> {
+        match (&self.color, &self.compressor) {
+            (DicomColorType::Gray16(_), Compressor::Uncompressed(c)) => {
+                <TiffSaver as SaveToTiff<Gray16, Uncompressed>>::save(
+                    self, encoder, image, metadata, *c,
+                )
+            }
+            (DicomColorType::Gray16(_), Compressor::Packbits(c)) => {
+                <TiffSaver as SaveToTiff<Gray16, Packbits>>::save(
+                    self, encoder, image, metadata, *c,
+                )
+            }
+            (DicomColorType::Gray16(_), Compressor::Lzw(c)) => {
+                <TiffSaver as SaveToTiff<Gray16, Lzw>>::save(self, encoder, image, metadata, *c)
+            }
+            (DicomColorType::Gray16(_), Compressor::Deflate(c)) => {
+                <TiffSaver as SaveToTiff<Gray16, Deflate>>::save(self, encoder, image, metadata, *c)
+            }
+            (DicomColorType::RGB8(_), Compressor::Uncompressed(c)) => {
+                <TiffSaver as SaveToTiff<RGB8, Uncompressed>>::save(
+                    self, encoder, image, metadata, *c,
+                )
+            }
+            (DicomColorType::RGB8(_), Compressor::Packbits(c)) => {
+                <TiffSaver as SaveToTiff<RGB8, Packbits>>::save(self, encoder, image, metadata, *c)
+            }
+            (DicomColorType::RGB8(_), Compressor::Lzw(c)) => {
+                <TiffSaver as SaveToTiff<RGB8, Lzw>>::save(self, encoder, image, metadata, *c)
+            }
+            (DicomColorType::RGB8(_), Compressor::Deflate(c)) => {
+                <TiffSaver as SaveToTiff<RGB8, Deflate>>::save(self, encoder, image, metadata, *c)
+            }
+        }
+    }
+
+    pub fn save_all(
```

Comment:
Can we leverage some of the similar code under `pub trait SaveToTiff<C, D>...fn save_all(` (line 64)?

Reply:
I couldn't do this directly without a large match statement for template dispatching. I went ahead and removed `save_all()` since it's trivial to do `images.into_iter().try_for_each(...)`. I also added a convenience method for opening TIFFs for writing